File: ios/rnProject/video/ZWStreamPlayer.m
//
//  ZWStreamPlayer.m
//  VideoEx1
//
//  Created by 改革丰富 on 2018/7/26.
//  Copyright © 2018年 改革丰富. All rights reserved.
//

#include <sys/types.h>
#import "ZWStreamPlayer.h"
#import "ZWHardDecoder.h"
#import "ZWAudioUnit.h"
#import "CommonPublic.h"
#import "MediaStreamServerC.h"
#import "ZWRtp808Frame.h"


#define logd(...) LogMessageA(0, "KEY", _channel, __VA_ARGS__)
#define logi(...) LogMessageA(1, "KEY", _channel, __VA_ARGS__)
#define logw(...) LogMessageA(2, "KEY", _channel, __VA_ARGS__)
#define loge(...) LogMessageA(3, "KEY", _channel, __VA_ARGS__)
int OnDataBlockFunc(const uint8_t* buffer, int bufferLength, void* data);
#define VIDEO_BUFFER_MAX 1024*1024*15

static const int PAYLOAD_TYPE_MASK = ((1 << 7) - 1);

@interface ZWFrameData : NSObject

@property (nonatomic, assign) uint8_t* data;
@property (nonatomic, assign) uint32_t trueLen;
@property (nonatomic, assign) int type;
@property (nonatomic, assign) uint32_t duration;
@property (nonatomic, assign) uint64_t startTime;

@end

@implementation ZWFrameData

-(void)dealloc
{
    if (_data) {
        free(_data);
        _data = nil;
    }
}

@end

@interface ZWStreamPlayer()

@property (nonatomic, strong) ZWStreamView* view;
@property (nonatomic, strong) ZWHardDecoder* decoder;
@property (nonatomic, strong) NSMutableArray<ZWFrameData*>* frameDataArray;
@property (nonatomic, strong) NSMutableArray<ZWFrameData*>* audioDataArray;
@property (nonatomic, strong) dispatch_queue_t queue;
@property (nonatomic, strong) dispatch_queue_t audioQueue;
@property (nonatomic, strong) dispatch_semaphore_t videoSem;
@property (nonatomic, strong) dispatch_semaphore_t audioSem;
@property (nonatomic, strong) dispatch_semaphore_t waitSem;
@property (nonatomic, assign) uint64_t lastTime;
@property (nonatomic, assign) uint64_t currentTime;
@property (nonatomic, assign) uint64_t timeBase;
@property (nonatomic, assign) int timeAdd;
@property (atomic, assign) int isStart;
@property (nonatomic, assign) size_t allLen;
@property (nonatomic, assign) int sampleRate;
@property (nonatomic, assign) bool isPlayAudio;

@property (nonatomic, assign) int isRender;

@end

@implementation ZWStreamPlayer
{
    void* _stream;
    int _channel;
    NSString* _playType;
    bool _skipFrames;
    char _message[1024];
    
    int _width;
    int _height;
}
-(instancetype)initWithUri:(NSString*)uri view:(ZWStreamView*)view delegate:(id<ZWStreamPlayerDelegate>)delegate sampleRate:(int)sampleRate enableAudio:(bool)enableAudio  withChannel:(int)channel withPlayType:(NSString*)playType
{
    self = [super init];
    if(self)
    {
        _uri = uri;
        _view = view;
        _isRender = 0;
        self.delegate = delegate;
        _decoder = [[ZWHardDecoder alloc] init];
        _queue = dispatch_queue_create([uri UTF8String], DISPATCH_QUEUE_SERIAL);
        _audioQueue = dispatch_queue_create("audio", DISPATCH_QUEUE_SERIAL);
        _frameDataArray = [[NSMutableArray alloc] init];
        _audioDataArray = [[NSMutableArray alloc] init];
        _videoSem = dispatch_semaphore_create(1);
        _audioSem = dispatch_semaphore_create(1);
        _waitSem = dispatch_semaphore_create(0);
        _isStart = 1;
        _sampleRate = sampleRate;
        _isPlayAudio = enableAudio;
        _channel = channel;
        _playType = playType;
        _skipFrames = ![_playType isEqual:@"PlayBack"];
        bool bret = [self startTask:uri];
        if (!bret) {
            return nil;
        }
        [self timeLoop];
    }
    return self;
}

-(void)dealloc
{
    logd("destroy stream player.");
}

-(void)playAudio
{
    _isPlayAudio = true;
}

-(void)closeAudio
{
    _isPlayAudio = false;
}-(void)close
{
    logd("close - stopServer");
    if (_stream) {
        void* freestream = _stream;
        _stream = nil;
        stopServer(freestream);
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            while(freestream && checkServerRunning(freestream)) {
                usleep(1000*100);
                freeServer(&freestream);
            }
        });
    }
    dispatch_semaphore_wait(_videoSem, DISPATCH_TIME_FOREVER);
    dispatch_semaphore_wait(_audioSem, DISPATCH_TIME_FOREVER);
    self.isStart = 0;
    dispatch_semaphore_signal(_videoSem);
    dispatch_semaphore_signal(_audioSem);
    dispatch_sync(_queue, ^{});
    dispatch_sync(_audioQueue, ^{});
    if(self.delegate)
    {
        [self.delegate onState:4];
    }
}
-(void)sendMessage:(NSString*)message
{
    logd("----testplay-send message(%d):%s",message.length, message.UTF8String);
    if (_stream&&checkServerRunning(_stream)) {
//        NSData* data = [ZWRtp808Frame encodeMessage:message];
        sendRtpMessage(_stream, message.UTF8String, (int)message.length);
        cleanBuffer(_stream);
        [self clearFrameData];
    } else {
        loge("error sendMessage");
    }
}

-(bool)startTask:(NSString*)uri
{
    logd("ZWStreamPlayer::startTask - start(%s)",uri?uri.UTF8String:"");
    do {
        if (!uri) {
            break;
        }
        _stream = initServer(uri.UTF8String, _channel, OnDataBlockFunc, (__bridge void *)(self));
        if (!_stream) {
            break;
        }
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            bool bret = runServer(_stream);
            if(!bret)
            {
                logd("startTask - stopServer");
                if(self.delegate)
                {
                    [self.delegate onState:1];
                }
                stopServer(_stream);
                return;
            }
            if(self.delegate)
            {
                [self.delegate onState:0];
            }
            while(_stream&&checkServerRunning(_stream)) {
                //                logd("Heartbeat");
                usleep(5000000);
                sendHeartbeat(_stream);
            }
            logd("Heartbeat - over;_stream=%p;running=%d",_stream,checkServerRunning(_stream));
            if(self.delegate)
            {
                [self.delegate onState:2];
            }
        });
        return true;
    } while (false);
    loge("fail url:%s",uri?uri.UTF8String:"null");
//    if(self.delegate)
//    {
//        [self.delegate onState:1];
//    }
    return false;
}
-(void)timeLoop
{
    __weak ZWStreamPlayer* weakSelf = self;
    dispatch_async(_queue, ^{
        __strong ZWStreamPlayer* player = weakSelf;
        while(player.isStart)
        {
            dispatch_semaphore_wait(player.videoSem, DISPATCH_TIME_FOREVER);
            if(player.frameDataArray.count>0)
            {
                ZWFrameData* frameData = player.frameDataArray.firstObject;
                if(frameData)
                {
                    dispatch_semaphore_signal(player.videoSem);
                    uint64_t time = (uint64_t) ([[NSProcessInfo processInfo] systemUptime] * 1000000);
                    if (frameData.startTime>time) {
                        int64_t sleepTime = frameData.startTime - time;
                        if(sleepTime > 0)
                        {
                            usleep((uint32_t)sleepTime);
                        }
                    }
                    // NSLog(@"play video data");
                    CVPixelBufferRef pixelBuffer = [player.decoder decode:frameData.data
                                                                      len:frameData.trueLen
                                                                frameType:frameData.type];
                    if(pixelBuffer)
                    {
                        [player.view draw:pixelBuffer];
                        if(self.delegate)
                        {
                            int width = (int)CVPixelBufferGetWidth(pixelBuffer);
                            int height = (int)CVPixelBufferGetHeight(pixelBuffer);
                            if (_width!=width||_height!=height) {
                                _width = width;
                                _height=height;
                                [self.delegate onVideoSize:_height height:_height];
                                logd("onVideoSize:%d-%d",_width,_height);
                            }
                        }
                        CFRelease(pixelBuffer);
                        if(!player.isRender)
                        {
                            player.isRender = 1;
                            if(player.delegate)
                            {
                                [player.delegate onState:3];
                            }
                        }
                    }
                    dispatch_semaphore_wait(player.videoSem, DISPATCH_TIME_FOREVER);
                    if (player.frameDataArray.count>0) {
                        
                        [player.frameDataArray removeObjectAtIndex:0];
                        player.allLen -= frameData.trueLen;
                    }
                }
                
            }
            else
            {
                dispatch_semaphore_wait(player.waitSem, dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_MSEC));
            }
            dispatch_semaphore_signal(player.videoSem);
        }
        player = nil;
    });
    
    dispatch_async(_audioQueue, ^{
        __strong ZWStreamPlayer* player = weakSelf;
        ZWAudioUnit *audioUnit = [[ZWAudioUnit alloc] initAudioUnit:player.sampleRate];
        while (player.isStart) {
            @autoreleasepool {
                dispatch_semaphore_wait(player.audioSem, DISPATCH_TIME_FOREVER);
                if (player.audioDataArray.count) {
                    ZWFrameData* frameData = player.audioDataArray.firstObject;
                    if (frameData == nil)
                    {
                        continue;
                    }
                    uint8_t *data = frameData.data + 26;
                    // NSLog(@"play audio data");
                    [audioUnit play:data dataLen:frameData.trueLen audioFormat:frameData.type];
                    [player.audioDataArray removeObjectAtIndex:0];
                    dispatch_semaphore_signal(player.audioSem);
                    continue;
                }
                dispatch_semaphore_wait(player.waitSem, dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_MSEC));
                dispatch_semaphore_signal(player.audioSem);
            }
        }
        [audioUnit stop];
        [audioUnit audio_release];
    });
}

-(void)pushFrameData:(ZWFrameData*)frameData
{
    dispatch_semaphore_wait(_videoSem, DISPATCH_TIME_FOREVER);
    _allLen += frameData.trueLen;
    _currentTime = _allLen > 30000000 ? 0 : _currentTime;
    [_frameDataArray addObject:frameData];
    if(_frameDataArray.count > 20)
    {
        _timeAdd = 2;
    }
    else if (_frameDataArray.count < 10)
    {
        _timeAdd = 0;
    }
    frameData.startTime -= _timeAdd;
    dispatch_semaphore_signal(_videoSem);
}
-(void)pushAudioData:(ZWFrameData*)frameData {
    dispatch_semaphore_wait(_audioSem, DISPATCH_TIME_FOREVER);
    [_audioDataArray addObject:frameData];
    dispatch_semaphore_signal(_audioSem);
}
-(void)clearFrameData{
    {
        dispatch_semaphore_wait(_videoSem, DISPATCH_TIME_FOREVER);
        [_frameDataArray removeAllObjects];
        _allLen = 0;
        _timeBase=0;
        _currentTime=0;
        logd("clean frame data:%d",_allLen);
        dispatch_semaphore_signal(_videoSem);
    }
    {
        dispatch_semaphore_wait(_audioSem, DISPATCH_TIME_FOREVER);
        [_audioDataArray removeAllObjects];
        dispatch_semaphore_signal(_audioSem);
    }
}

-(int)saveData:(const uint8_t*)buffer len:(int)len time:(uint64_t)time
{
    //    logd("ZWStreamPlayer::saveData - start");
    int proDataLen = -1;
    const uint8_t* dataSource = buffer;
    int type = (dataSource[15] & 0xf0) >> 4;
    if(type == 0 || (type < 3 && _allLen < VIDEO_BUFFER_MAX))
    {
        if (_allLen>VIDEO_BUFFER_MAX) {
            if (_skipFrames) {
                [self clearFrameData];
            } else {
                logd("VIDEO_BUFFER_MAX=%d,_allLen=%d",VIDEO_BUFFER_MAX,_allLen);
                return 0;
            }
        }
        int frameHeadLen = 30;
        uint64_t currentTime = ntohll(*((uint64_t*)(dataSource + 16)));
        uint16_t duration = ntohs(*((uint16_t*)(dataSource + 26)));
        duration = (uint16_t) (_currentTime == 0 ? 0 : (duration > 0 ? duration : currentTime - _currentTime));
        _currentTime = currentTime;
        duration = (uint16_t) ((duration > 0 && duration < 500) ? duration : 40);
        int offset = 0;
        int trueLen = 0;
        int rtpLen = 0;
        //        logd("saveData - malloc");
        uint8_t* videoData = (uint8_t*)malloc(len);
        if (!videoData) {
            loge("ZWStreamPlayer saveData MALLOC ERROR! length=%d",len);
            return 1;
        }
        while(rtpLen <= 950 && offset < len)
        {
            rtpLen = ntohs(*((uint16_t*)(dataSource + offset + 28)));
            if (rtpLen+offset+frameHeadLen>len) {
                loge("rtp frame length error:rtpLen=%d,offset=%d,frameHeadLen=%d,len=%d",rtpLen,offset,frameHeadLen,len);
                break;
            }
            memcpy(videoData+trueLen, dataSource + offset + frameHeadLen, rtpLen);
            trueLen += rtpLen;
            offset += rtpLen + frameHeadLen;
        }
        proDataLen = 1;
        if(trueLen>0&&trueLen<len)
        {
            ZWFrameData* saveData = [[ZWFrameData alloc] init];
            saveData.data = videoData;
            saveData.trueLen = (uint32_t) trueLen;
            saveData.type = type;
            saveData.duration = (uint32_t) (duration * 1000);
            _timeBase = _timeBase == 0 ? time : _timeBase;
            saveData.startTime = _timeBase;
            _timeBase += saveData.duration;
            [self pushFrameData:saveData];
        } else {
            free(videoData);
            loge("rtp frame length error:trueLen=%d, len=%d",trueLen, len);
        }
    } else if (type == 3) {
        if (!_isPlayAudio) {
            return 1;
        }
//        logd("saveData - malloc audio");
        uint8_t* audioData = (uint8_t*)malloc(len);
        if (!audioData) {
            loge("ZWStreamPlayer saveData audio MALLOC ERROR!%d",len);
            return 1;
        }
        memcpy(audioData, dataSource, len);
        uint8_t payloadType = (uint8_t) (*(dataSource + 5) & PAYLOAD_TYPE_MASK);
        ZWFrameData* saveData = [[ZWFrameData alloc] init];
        saveData.data = audioData;
        saveData.type = payloadType;
        saveData.trueLen = ntohs(*((uint16_t*)(audioData + 24)));
        saveData.startTime = time;
        proDataLen = len;
        [self pushAudioData:saveData];
    } else if (type == 4) {
        if(self.delegate)
        {
            memset(_message, 0, 1024);
            memcpy(_message, dataSource+sizeof(struct Rtp808HeadBaseT), len - sizeof(struct Rtp808HeadBaseT));
            NSString* strMessage = [NSString stringWithUTF8String:_message];
            [self.delegate onMessage:strMessage];
        }
        proDataLen = 1;
    } else {
        proDataLen = 1;
    }
    return 1;
}

@end

int OnDataBlockFunc(const uint8_t* buffer, int bufferLength, void* data){
    uint64_t time = (uint64_t) ([[NSProcessInfo processInfo] systemUptime] * 1000000);
    if (data) {
        ZWStreamPlayer* pPlayer = (__bridge ZWStreamPlayer*)data;
        int iret = [pPlayer saveData:buffer len:bufferLength time:time];
        return iret;
    }
    return 1;
}

File: android/app/src/main/cpp/ZWStreamPlayer.cpp
//
// Created by Administrator on 2018/8/14.
//

#include "ZWStreamPlayer.h"
#include "ZWHardDecoder.h"
#include "ZWStreamGLView.h"
#include "Rtp808Frame.h"

#define logd(...) LogMessageA(0, _task.key.c_str(), _task.channel, __VA_ARGS__)
#define logi(...) LogMessageA(1, _task.key.c_str(), _task.channel, __VA_ARGS__)
#define logw(...) LogMessageA(2, _task.key.c_str(), _task.channel, __VA_ARGS__)
#define loge(...) LogMessageA(3, _task.key.c_str(), _task.channel, __VA_ARGS__)

ZWStreamPlayer::ZWStreamPlayer(int sampleRate, int sampleBufSize, bool enableAudio)
        : _channel(0), _sampleRate(sampleRate),
          _sampleBufSize(sampleBufSize), _enableAudio(enableAudio) {

#if ENABLE_RTPFRAME_LIST
    _lstFrames = std::make_shared<Rtp808FrameList>();
//        _lstFrames->autoLostFrame(true);
#endif
}

ZWStreamPlayer::~ZWStreamPlayer()
{
    stop();
    delete _webSocketStateDelegate;
}

void ZWStreamPlayer::pushFrameData(std::shared_ptr<ZWFrameData> frameData) {
    if(_frameDataList.size() > 19)
    {
        frameData -> startTime -= 2;
    }
    {
        std::lock_guard<std::mutex> lock(_mutex);
        _allLen += frameData -> dataLen;
        _frameDataList.push_back(frameData);
//        logd("pushFrameData");
    }
    _condition.notify_one();
}

void ZWStreamPlayer::pushAudioData(std::shared_ptr<ZWFrameData> frameData) {
    {
        std::unique_lock<std::mutex> lock(_audioMutex);
        _audioDataList.push_back(frameData);
    }
    _audioCondition.notify_one();
}
void ZWStreamPlayer::clearFrameData() {
    {
        std::lock_guard<std::mutex> lock(_mutex);
        _allLen = 0;
        _timeBase = 0;
        _currentTime = 0;
        _frameDataList.clear();
        logi("clearFrameData _allLen=%d",_allLen);
    }
    {
        std::lock_guard<std::mutex> lock(_audioMutex);
        _audioDataList.clear();
    }
}

int ZWStreamPlayer::saveData(const uint8_t* buffer, int bufferLengh, uint64_t time)
{
    if (_logReadFlag > 150)
    {
//        logd("ZWStreamPlayer::saveData %s-%d data : %d, _allLen=%d,list=%d", _task.key.c_str(),_channel, bufferLengh,_allLen,_frameDataList.size());
        if (!_audioMessage.empty()){
            loge("%s",_audioMessage.c_str());
        }
        _logReadFlag=0;
    }
    else
    {
        _logReadFlag++;
    }

    const uint8_t* dataSource = buffer;
    uint8_t bitData = ((uint8_t)(dataSource[15] & 0xf0u));
    int packageNumber = (dataSource[6]<<8 | dataSource[7]);
//    logd(" package number=%d",packageNumber);
    int type =  bitData >> 4u; // data type: 0 video I frame; 1 video P frame; 2 video B frame; 3 audio frame
//        if(((_currentTime != 0 && type < 3) || type == 0) && ((_currentTime == 0 && _allLen < 10000000) || (_currentTime != 0 && _allLen < 30000000))) {
    if(type == 0 || (_allLen < VIDEO_BUFFER_MAX && type < 3)) {
//        return 1;
        // 关键帧 或 缓存足够时
        if (_allLen>=VIDEO_BUFFER_MAX){
            if(_skipFrames){
                clearFrameData();
            } else {
                // 阻塞
//                logd("_skipFrames=%d",_skipFrames);
                return 0;
            }
        }
        _showLoop = false;
        int frameHeadLen = 30;
        uint64_t currentTime = htonq(*((uint64_t *) (dataSource + 16)));
        uint16_t duration = ntohs(*((uint16_t *) (dataSource + 26)));
        duration = static_cast<uint16_t>(_currentTime == 0 ? 0 : (duration > 0 ? duration : currentTime - _currentTime));
        _currentTime = currentTime;
        duration = static_cast<uint16_t>((duration > 0 && duration < 500) ? duration : 40);


        auto videoData = (uint8_t*)malloc(bufferLengh);
        if (!videoData){
            loge("ZWStreamPlayer::saveData MALLOC ERROR！%d", bufferLengh);
            return 1;
        }
//        std::vector<uint8_t> videoData;
//        videoData.resize(bufferLengh);
        int offset = 0;
        int videoDataLen = 0;
        while(bufferLengh - offset>0)
        {
            if (bufferLengh - offset < frameHeadLen)
            {
                logd("bufferLengh - offset=%d,frameHeadLen=%d",bufferLengh - offset,frameHeadLen);
                break;
            }
            int rtpVideoDataLen = ntohs(*((uint16_t *) (dataSource + offset + 28)));
            if (bufferLengh - offset - frameHeadLen < rtpVideoDataLen)
            {
                logd("bufferLengh - offset - frameHeadLen=%d,rtpVideoDataLen=%d",bufferLengh - offset - frameHeadLen,rtpVideoDataLen);
                break;
            }
            memcpy(videoData+videoDataLen, dataSource+offset+frameHeadLen, rtpVideoDataLen);
            videoDataLen+=rtpVideoDataLen;
            offset+=frameHeadLen+rtpVideoDataLen;
        }
//        offset = 0;
        if(offset>0){
//            logd("offset>0");
            std::shared_ptr<ZWFrameData> saveData = std::make_shared<ZWFrameData>();
            saveData->data = videoData;
            saveData->dataLen = videoDataLen;
            saveData->type = type;
            saveData->duration = static_cast<uint32_t>(duration * 1000);
            _timeBase = _timeBase == 0 ? time : _timeBase;
            saveData->startTime = _timeBase;
            _timeBase += saveData->duration;
            pushFrameData((saveData));
        } else {
            free(videoData);
        }
    } else if(type == 3) {
//        logd("ZWStreamPlayer::saveData: _enableAudio=%d, ",_enableAudio);
//        _enableAudio = false;
        if (!_enableAudio)
            return 1;
        int frameHeadLen = 26;
//        std::vector<uint8_t> audioData;
        auto audioData = (uint8_t*)malloc(bufferLengh);
        if (!audioData){
            loge("ZWStreamPlayer::saveData audio MALLOC ERROR！%d", bufferLengh);
            return 1;
        }
        int offset = 0;
        int audioDataLength = 0;
        while(true)
        {
            if (bufferLengh - offset < frameHeadLen)
            {
                break;
            }
            int rtpDataLen = ntohs(*((uint16_t *) (dataSource + offset + 24)));
            if (bufferLengh - offset - frameHeadLen < rtpDataLen)
            {
                break;
            }
            memcpy(audioData, dataSource+offset+frameHeadLen, rtpDataLen);

            offset+=frameHeadLen+rtpDataLen;
            audioDataLength+=rtpDataLen;
//            logd("one audio frame:%d",frameHeadLen+rtpDataLen);
        }
//        offset = 0;
        if (offset>0){
            uint8_t payloadType = *(dataSource + 5) & PAYLOAD_TYPE_MASK;
            std::shared_ptr<ZWFrameData> saveData = std::make_shared<ZWFrameData>();
            saveData->data = audioData;
            saveData->dataLen = audioDataLength;
            saveData->payloadType = payloadType;
            saveData->startTime = time;
            pushAudioData((saveData));
        } else {
            free(audioData);
        }
    } else if(EDataType::TFRAME == type){
        std::string message(bufferLengh+1,'\0');
        if (buffer[5] == 0xf0){
            logd("error message");
        } else {
            memcpy(message.data(), buffer+sizeof(Rtp808HeadBaseT), bufferLengh-sizeof(Rtp808HeadBaseT));
            logd("onMessage:%s,packageNumber=%d",message.c_str(),packageNumber);
            _webSocketStateDelegate->onMessage(message);
        }
    } else {
        if(_skipFrames){
            _showLoop = true;
            logd("ZWStreamPlayer::startTimeLoop %s-%d, _allLen=%d,list=%d,type=%d", _task.key.c_str(),_channel, _allLen,_frameDataList.size(),type);
        }
//        logd("ZWStreamPlayer::saveData  %s-%d, unknow type:%d,_currentTime:%lld,_allLen:%d, _frameDataList:%d", _task.key.c_str(),_channel, type,_currentTime,_allLen,_frameDataList.size());
    }
    return 1;
}

void ZWStreamPlayer::initTaskInfo(TaskInfoPar& task, const std::string& uri)
{
    logd(("initTaskInfo _cacheDir="+_cacheDir).c_str());
    // "ws://58.87.111.217:7971/4ae969bb-b90f-4a65-9d6a-99dd85a5367f"
    task.ip = "";
    task.port = 0;
    task.key = "";
    task.channel = _channel;
    task.cacheDir = _cacheDir;
    do
    {
        std::string strPip = "ws://";
        std::string strPport = ":";
        std::string strPkey = "/";
        auto pip = uri.find(strPip);
        if (pip == std::string::npos)
        {
            break;
        }
        auto pport = uri.find(strPport, pip+strPip.size());
        if (pport == std::string::npos)
        {
            break;
        }
        auto pkey = uri.find(strPkey, pport+strPport.size());
        if (pkey == std::string::npos)
        {
            break;
        }
        task.ip = uri.substr(pip+strPip.size(),pport-pip-strPip.size());
        task.port = atoi(uri.substr(pport+strPport.size(),pkey-pport-strPport.size()).c_str());
        task.key = uri.substr(pkey+strPkey.size());
    }while(false);
    logd("initTaskInfo:ip=%s,port=%d,key=%s,channel=%d",task.ip.c_str(),task.port,task.key.c_str(),task.channel);
}
void ZWStreamPlayer::startWebSocket(const char *uri)
{
    logd("startWebSocket:%s",uri);
    _websocketThread = std::make_unique<std::thread>([this, uri](){

        initTaskInfo(_task, uri);
//        _task.port = 8971;
        if(_isStart)
        {
//            _webSocketStateDelegate->onState(WebSocketStateDelegate::CONNECTED);
            _task.cb = [this](const uint8_t* buffer, int bufferLengh, void* data) {
//                return 1;
                uint64_t time = getCurrentTime();
                return saveData(buffer, bufferLengh, time);
            };
            if (_stream) {
                _stream->StopTask();
            }
            _stream = std::make_shared<MediaStreamServer>();
            auto bret = _stream->SetupTask(_task);
            auto retstate = bret?WebSocketStateDelegate::CONNECTED:WebSocketStateDelegate::ERROR;
            _webSocketStateDelegate->onState(retstate);
            if (bret) {
                int sendOffset = 5000000;
                int lastOffset = 0;
                int stepSleep = 100000;
                while(_stream->Running()) {
                    usleep(stepSleep);
                    lastOffset += stepSleep;
                    if (lastOffset>=sendOffset){
                        _stream->Send((const uint8_t*)"0", 1);
                        lastOffset = 0;
                    }
                }
                logd("startWebSocket - stream end");
            }
        }
        //close
        _webSocketStateDelegate->onState(WebSocketStateDelegate::CLOSED);
    });
}
void ZWStreamPlayer::startTimeLoop() {
    logd("startTimeLoop");
    startVideoThread();
    startAudioThread();
}

void ZWStreamPlayer::startVideoThread() {
    _renderThread = std::make_unique<std::thread>([this]() {
        pthread_setname_np(pthread_self(), "ZW-VideoRender");
        _videoWidth = 0;
        _videoHeight = 0;
#if DECODE_JAVA

#elif DECODE_2_SURFACE
        auto callback = [this](uint8_t *data, int width, int height, int type) {
            _webSocketStateDelegate->onState(WebSocketStateDelegate::VIDEO_OPEN);
            if (_videoWidth!=width || _videoHeight!=height)
            {
                logd("_videoWidth=%d,_videoHeight=%d,width=%d,height=%d",_videoWidth,_videoHeight,width,height);
                _videoWidth = width;
                _videoHeight = height;
                _webSocketStateDelegate->onVideoSize(_videoWidth, _videoHeight);
            }

        };
        std::shared_ptr<ZWHardDecoder> videoDecoder = std::make_shared<ZWHardDecoder>(callback, _nativeWindow);
#else
        std::shared_ptr<ZWStreamGLView> view(ZWStreamGLView::create(_nativeWindow, viewWidth, viewHeight));
        auto callback = [this, view](uint8_t *data, int width, int height, int type) {
            _webSocketStateDelegate->onState(WebSocketStateDelegate::VIDEO_OPEN);
            if (_videoWidth!=width || _videoHeight!=height)
            {
                logd("_videoWidth=%d,_videoHeight=%d,width=%d,height=%d",_videoWidth,_videoHeight,width,height);
                _videoWidth = width;
                _videoHeight = height;
                _webSocketStateDelegate->onVideoSize(_videoWidth, _videoHeight);
            }
            if (view) {
                view->setWindowsSize(_viewwidth, _viewheight);
                view->setProgram(type == 19 ? 2 : 1);
                view->draw(data, _videoWidth, _videoHeight);
            }
        };
        std::shared_ptr<ZWHardDecoder> videoDecoder = std::make_shared<ZWHardDecoder>(callback);
#endif
        std::chrono::seconds waitTime(2);
        while (_isStart) {
            usleep(1);
            {
                if (_showLoop){
                    logd("ZWStreamPlayer::startTimeLoop %s-%d, _allLen=%d,list=%d", _task.key.c_str(),_channel, _allLen,_frameDataList.size());
                }

                std::unique_lock<std::mutex> lock(_mutex);
                if (_frameDataList.empty()) {
                    _condition.wait_for(lock, waitTime);
                    continue;
                }

                auto frameData = _frameDataList.front();
                _frameDataList.pop_front();
                _allLen -= frameData->dataLen;
                lock.unlock();

                uint64_t time = getCurrentTime();
                if(frameData->startTime>time){
                    int64_t sleepTime = frameData->startTime - time;
                    if (sleepTime > 0) {
                        //logd("video thread sleep %lld;", sleepTime);
                        usleep((uint32_t) sleepTime);
                    }
                } else {
//                    int64_t sleepTime = time - frameData->startTime;
//                    logd("2video thread sleep %lld;", sleepTime);
                }

                // 解码缓存获取失败时，数据是否直接丢弃
#if DECODE_JAVA
//                logd("onVideoData");
                _webSocketStateDelegate->onVideoData((const char*)frameData->data, frameData->dataLen);
                _webSocketStateDelegate->onState(WebSocketStateDelegate::VIDEO_OPEN);
#else
                videoDecoder->decodeData(frameData->data, frameData->dataLen);
#endif
            }
        }
    });
}

void ZWStreamPlayer::startAudioThread() {
    _audioThread = std::make_unique<std::thread>([this]() {
        pthread_setname_np(pthread_self(), "ZW-AudioRender");
        std::chrono::seconds waitTime(3);
        audioPlayer = std::make_shared<ZWAudioPlayer>(_sampleRate, _sampleBufSize);
        if (_enableAudio) {
            audioPlayer->play();
        }
        while (_isStart) {
            usleep(1);
            {
                std::unique_lock<std::mutex> audioLock(_audioMutex);
                if (_audioDataList.empty()) {
                    _audioCondition.wait_for(audioLock, waitTime);
                    continue;
                }
                auto frameData = (_audioDataList.front());
                _audioDataList.pop_front();
                audioLock.unlock();

                uint64_t time = getCurrentTime();
                int64_t sleepTime = frameData->startTime - time;
                if (sleepTime > 0) {
                    usleep((uint32_t) sleepTime);
                }

                if (_enableAudio) {
                    // RTP Header's size for audio is 26 bytes, and pcm header's size is 8 bytes
                    std::string message;
                    auto audioRet = audioPlayer->store(frameData->payloadType, frameData->data, static_cast<int>(frameData->dataLen), message);
//                    auto audioRet = audioPlayer->store(frameData->payloadType, frameData->data + 26, static_cast<int>(frameData->dataLen - 26), message);
                    if (!audioRet){
                        _audioMessage = message;
                    } else {
                        _audioMessage = "";
                    }
                }
            }
        }
        audioPlayer = nullptr;
    });
}

void ZWStreamPlayer::play(const char* uri) {
    if (_isStart){
        logd("Already playing!");
        return;
    }
    _isStart = true;
    if(_playType == "PlayBack") {
        _skipFrames=false;
    }
    logi("play %s", _playType.c_str());
    startWebSocket(uri);
    startTimeLoop();
}

void ZWStreamPlayer::playAudio() {
    if (!_isStart) {
        return;
    }
    if (audioPlayer == nullptr) {
        return;
    }
    audioPlayer->play();
    _enableAudio= true;
}

void ZWStreamPlayer::stopAudio() {
    if (!_isStart && audioPlayer != nullptr) {
        return;
    }
    _enableAudio = false;
    _audioCondition.notify_one();
    if (audioPlayer == nullptr) {
        return;
    }
    audioPlayer->stop();
}
void ZWStreamPlayer::sendMessage(const std::string& msg){
    if(_stream && _stream->Running())
    {
        logi("ZWStreamPlayer - sendMessage:%s",msg.c_str());
//        auto ptr = Rtp808Frame::encodeMessage(msg, size);
        _stream->Send((const uint8_t*)msg.data(),msg.size());
        _stream->Clean();
        clearFrameData();
    }
}

void ZWStreamPlayer::stop() {
    logi("ZWStreamPlayer - stop.");
    {
        std::unique_lock<std::mutex> lock(_mutex);
        if (!_isStart) {
            _webSocketStateDelegate->onState(WebSocketStateDelegate::VIDEO_CLOSED);
            logw("ZWStreamPlayer - stop: 未开始播放 .");
            return;
        }

        if(_stream && _stream->Running())
        {
            _stream->StopTask();
        }
        _isStart = false;
    }
    logi("ZWStreamPlayer - play stop wait.");
    _condition.notify_one();
    _audioCondition.notify_one();
    _renderThread->join();
    _audioThread->join();
    _websocketThread->join();

    logi("ZWStreamPlayer - play stop state.");
    _webSocketStateDelegate->onState(WebSocketStateDelegate::VIDEO_CLOSED);
}
